File: secretspec-rb/lib/secretspec.rb
# frozen_string_literal: true

# Ruby SDK for SecretSpec, a declarative secrets manager.
#
# A thin client over the secretspec-ffi C ABI. The Rust resolver is statically
# linked into a native extension (secretspec_ext), so the SDK inherits every
# provider with no Ruby-side logic and there is nothing to locate at runtime.
# Mirrors the Rust derive crate's vocabulary.

require "json"

# The compiled extension lives next to this file in a source/dev checkout, but in
# an installed gem RubyGems places it in a separate extensions dir already on
# $LOAD_PATH. Put this file's dir on the path so the absolute require resolves in
# both layouts.
$LOAD_PATH.unshift(__dir__) unless $LOAD_PATH.include?(__dir__)
require "secretspec/secretspec_ext"

module Secretspec
  # Response wire-format version this SDK understands. Tracks secretspec-ffi's
  # RESOLVE_SCHEMA_VERSION; a mismatch means the loaded library is incompatible.
  RESOLVE_SCHEMA_VERSION = 1

  # Wire-format version of the value-free report. Tracks secretspec's
  # RESOLUTION_REPORT_SCHEMA_VERSION.
  REPORT_SCHEMA_VERSION = 1

  # A resolution failure (bad manifest, provider error, reason policy).
  class Error < StandardError
    attr_reader :kind

    def initialize(kind, message)
      @kind = kind
      super("#{message} (kind: #{kind})")
    end
  end

  # One or more required secrets were not found anywhere.
  class MissingRequiredError < Error
    attr_reader :missing

    def initialize(missing)
      @missing = missing
      super("missing_required", "missing required secret(s): #{missing.join(', ')}")
    end
  end

  # One resolved secret. Exactly one of +value+ / +path+ is set.
  ResolvedSecret = Struct.new(:value, :path, :as_path, :source, :source_provider) do
    # The usable string: the file path for as_path secrets, else the value.
    def get
      as_path ? path : value
    end
  end

  # A successful resolution, mirroring the Rust Resolved wrapper.
  Resolved = Struct.new(:provider, :profile, :secrets, :missing_optional) do
    # Export each resolved secret into ENV by its declared name. Secrets with no
    # usable value (e.g. under no_values) are skipped rather than deleted from
    # ENV (assigning nil would remove the variable).
    def set_as_env!
      secrets.each do |name, secret|
        value = secret.get
        ENV[name] = value unless value.nil?
      end
    end

    # Flat { "SECRET_NAME" => value } hash (the file path for as_path). A secret
    # with no usable value (e.g. under no_values) maps to nil, matching the null
    # the other SDKs emit. Feed this to a quicktype-generated deserializer (e.g.
    # from_dynamic!). See `secretspec schema`.
    def fields
      secrets.transform_values(&:get)
    end

    # Remove the temp files backing any as_path secrets in this result. The
    # resolver persists those files (mode 0400) so their paths stay valid after
    # resolve returns; the caller owns their lifetime. Call #close (or pass a
    # block to Builder#load, which closes automatically) when done so secret
    # files do not accumulate in the temp dir. A file already gone is not an
    # error.
    def close
      secrets.each_value do |secret|
        next unless secret.as_path && secret.path

        File.delete(secret.path) if File.exist?(secret.path)
      end
      nil
    end
  end

  # Value-free resolution outcome for one declared secret: how it would resolve
  # and from where, never the value itself.
  SecretReport = Struct.new(:name, :status, :required, :source_provider,
                            :default_applied, :generated, :as_path)

  # A value-free resolution snapshot. Unlike Resolved, a missing required secret
  # is a "missing_required" status here, not an error, so a report describes a
  # profile even when its secrets are not all available.
  Report = Struct.new(:provider, :profile, :secrets)

  # The narrow C ABI, statically linked into the secretspec_ext extension. The
  # Native.c_resolve / c_abi_version C functions are defined in
  # ext/secretspec/secretspec_ext.c; these wrappers add the Ruby-side error type.
  module Native
    class << self
      def resolve(request_json)
        result = c_resolve(request_json)
        raise Error.new("ffi", "secretspec_resolve returned null") if result.nil?

        result
      end

      def abi_version
        c_abi_version
      end
    end
  end

  # Entry point mirroring the derive crate's SecretSpec::builder().
  class SecretSpec
    def self.builder
      Builder.new
    end
  end

  # Fluent builder for a resolution.
  class Builder
    def initialize
      @request = {}
    end

    def with_path(path)
      @request["path"] = path if path
      self
    end

    def with_provider(provider)
      @request["provider"] = provider if provider
      self
    end

    def with_profile(profile)
      @request["profile"] = profile if profile
      self
    end

    def with_reason(reason)
      @request["reason"] = reason if reason
      self
    end

    # Omit secret values, returning only structure and provenance.
    def with_no_values(no_values = true)
      @request["no_values"] = no_values
      self
    end

    # Resolve the secrets. Raises MissingRequiredError if a required secret is
    # missing, and Error for any other failure.
    #
    # Without a block, returns the Resolved (the caller should #close it when
    # done to clean up any as_path temp files). With a block, yields the Resolved
    # and closes it afterwards, returning the block's value.
    def load
      response = parse_response(JSON.generate(@request), "resolve", RESOLVE_SCHEMA_VERSION)

      missing = response["missing_required"] || []
      raise MissingRequiredError.new(missing) unless missing.empty?

      secrets = {}
      (response["secrets"] || {}).each do |name, entry|
        secrets[name] = ResolvedSecret.new(
          entry["value"], entry["path"], entry["as_path"] || false,
          entry["source"], entry["source_provider"]
        )
      end

      resolved = Resolved.new(
        response["provider"], response["profile"], secrets,
        response["missing_optional"] || []
      )
      return resolved unless block_given?

      begin
        yield resolved
      ensure
        resolved.close
      end
    end

    # Resolve a value-free Report (the inventory/preflight view, the same one the
    # CLI exposes as `check --json`). Unlike #load, never raises
    # MissingRequiredError: a missing required secret appears as a SecretReport
    # with status "missing_required".
    def report
      request = @request.merge("mode" => "report")
      response = parse_response(JSON.generate(request), "report", REPORT_SCHEMA_VERSION)

      secrets = (response["secrets"] || []).map do |s|
        SecretReport.new(s["name"], s["status"], s["required"],
                         s["source_provider"], s["default_applied"],
                         s["generated"], s["as_path"])
      end
      Report.new(response["provider"], response["profile"], secrets)
    end

    private

    # Resolve a JSON request payload and return the validated "response" hash, or
    # raise. +kind+ is "resolve" or "report"; it selects the schema version to
    # enforce and labels the version-mismatch message.
    def parse_response(payload, kind, expected_version)
      envelope = JSON.parse(Native.resolve(payload))

      unless envelope["ok"]
        err = envelope["error"] || {}
        raise Error.new(err["kind"] || "unknown", err["message"] || "")
      end

      response = envelope["response"]
      raise Error.new("ffi", "secretspec_resolve reported ok with no response") if response.nil?

      version = response["schema_version"]
      unless version == expected_version
        raise Error.new("version",
                        "unsupported #{kind} schema version #{version} " \
                        "(expected #{expected_version}); the secretspec-ffi " \
                        "library and this SDK are out of sync")
      end

      response
    end
  end

  def self.abi_version
    Native.abi_version
  end
end

File: secretspec-go/secretspec.go
// Package secretspec is a Go SDK for SecretSpec, a declarative secrets manager.
//
// It is a thin client over the secretspec-ffi C ABI. Resolution (providers,
// chains, profiles, generation, as_path) happens entirely in the Rust core; this
// package marshals a JSON request to secretspec_resolve, parses the response
// envelope, and exposes it with the same vocabulary as the Rust derive crate.
//
// Two bindings select the native resolver at build time:
//   - default (no build tag): purego (dlopen, no cgo). The library is located via
//     the SECRETSPEC_FFI_LIB environment variable, an embedded copy, or a Cargo
//     target directory. This keeps `go get` toolchain-free.
//   - `-tags static`: cgo statically links libsecretspec_ffi.a, so the resolver
//     is embedded in the Go binary (fully static on Linux/musl). See README.
//
// Both bindings implement the same hooks (ensureLoaded, nativeResolve,
// nativeABIVersion); the code below is binding-agnostic.
package secretspec

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
)

// resolveSchemaVersion is the response wire-format version this SDK understands.
// It tracks secretspec-ffi's RESOLVE_SCHEMA_VERSION; a mismatch means the loaded
// library is incompatible with this SDK, so Load reports it rather than silently
// misparsing.
const resolveSchemaVersion = 1

// Error is a resolution failure (bad manifest, provider error, reason policy).
type Error struct {
	Kind    string
	Message string
}

func (e *Error) Error() string {
	return fmt.Sprintf("%s (kind: %s)", e.Message, e.Kind)
}

// MissingRequiredError reports required secrets that were not found anywhere.
type MissingRequiredError struct {
	Missing []string
}

func (e *MissingRequiredError) Error() string {
	return "missing required secret(s): " + strings.Join(e.Missing, ", ")
}

// ResolvedSecret is one resolved secret. Exactly one of Value / Path is set.
type ResolvedSecret struct {
	Value          *string
	Path           *string
	AsPath         bool
	Source         string
	SourceProvider *string
}

// Usable returns the secret's usable string and whether one is present: the file
// path for as_path secrets, otherwise the value. Both are absent when a
// value-less response (e.g. no_values) strips them, in which case ok is false.
// This is the null-aware accessor; the other SDKs express the same thing as a
// get() that returns null/None/nil.
func (s ResolvedSecret) Usable() (string, bool) {
	if s.AsPath {
		if s.Path != nil {
			return *s.Path, true
		}
		return "", false
	}
	if s.Value != nil {
		return *s.Value, true
	}
	return "", false
}

// Get returns the usable string: the file path for as_path secrets, else the
// value. It is the empty string when no usable value is present; use Usable to
// distinguish an absent value from a genuinely empty one.
func (s ResolvedSecret) Get() string {
	v, _ := s.Usable()
	return v
}

// Resolved is a successful resolution, mirroring the Rust Resolved wrapper.
type Resolved struct {
	Provider        string
	Profile         string
	Secrets         map[string]ResolvedSecret
	MissingOptional []string
}

// SetAsEnv exports each resolved secret into the process environment by name.
// Secrets with no usable value (e.g. under no_values) are skipped rather than
// exported as an empty string.
func (r *Resolved) SetAsEnv() error {
	for name, secret := range r.Secrets {
		if value, ok := secret.Usable(); ok {
			if err := os.Setenv(name, value); err != nil {
				return err
			}
		}
	}
	return nil
}

// Fields returns a flat map of SECRET_NAME -> value (the file path for as_path).
// A secret with no usable value (e.g. under no_values) maps to a nil pointer,
// which marshals to JSON null, matching the null the Python, Ruby, and Node SDKs
// emit; the value is a non-nil pointer otherwise.
func (r *Resolved) Fields() map[string]*string {
	out := make(map[string]*string, len(r.Secrets))
	for name, secret := range r.Secrets {
		if v, ok := secret.Usable(); ok {
			val := v
			out[name] = &val
		} else {
			out[name] = nil
		}
	}
	return out
}

// FieldsJSON marshals Fields() to JSON (a `{SECRET_NAME: value-or-null}` object),
// the input for a quicktype-generated deserializer (e.g. UnmarshalSecretSpec).
// See `secretspec schema`.
func (r *Resolved) FieldsJSON() ([]byte, error) {
	return json.Marshal(r.Fields())
}

// Close removes the temp files backing any as_path secrets in this result. The
// resolver persists those files (mode 0400) so their paths stay valid after
// resolve returns; the caller owns their lifetime. Call it (e.g.
// `defer resolved.Close()`) when done so secret files do not accumulate in the
// temp dir. Non-as_path secrets and a no_values result hold no path and are
// skipped, and a file already gone is not an error.
func (r *Resolved) Close() error {
	var firstErr error
	for _, secret := range r.Secrets {
		if secret.AsPath && secret.Path != nil {
			if err := os.Remove(*secret.Path); err != nil && !os.IsNotExist(err) && firstErr == nil {
				firstErr = err
			}
		}
	}
	return firstErr
}

// ABIVersion returns the version reported by the native resolver.
func ABIVersion() (string, error) {
	if err := ensureLoaded(); err != nil {
		return "", err
	}
	return nativeABIVersion()
}

// Builder configures a resolution, mirroring the derive crate's SecretSpec::builder().
type Builder struct {
	req map[string]any
}

// New starts a resolution builder.
func New() *Builder {
	return &Builder{req: map[string]any{}}
}

// set lazily initializes the request map so a zero-value Builder (e.g.
// `var b Builder` or `&Builder{}`, not just New()) does not panic with a
// nil-map write in the setters below.
func (b *Builder) set(key string, value any) *Builder {
	if b.req == nil {
		b.req = map[string]any{}
	}
	b.req[key] = value
	return b
}

func (b *Builder) WithPath(path string) *Builder     { return b.set("path", path) }
func (b *Builder) WithProvider(p string) *Builder    { return b.set("provider", p) }
func (b *Builder) WithProfile(p string) *Builder     { return b.set("profile", p) }
func (b *Builder) WithReason(reason string) *Builder { return b.set("reason", reason) }
func (b *Builder) WithNoValues(v bool) *Builder      { return b.set("no_values", v) }

// envelope is the response wrapper shared by the resolve and report paths,
// generic over its inner response type R.
type envelope[R any] struct {
	OK       bool       `json:"ok"`
	Response *R         `json:"response"`
	Error    *errorJSON `json:"error"`
}

type errorJSON struct {
	Kind    string `json:"kind"`
	Message string `json:"message"`
}

type secretJSON struct {
	Value          *string `json:"value"`
	Path           *string `json:"path"`
	AsPath         bool    `json:"as_path"`
	Source         string  `json:"source"`
	SourceProvider *string `json:"source_provider"`
}

type responseJSON struct {
	SchemaVersion   int                   `json:"schema_version"`
	Provider        string                `json:"provider"`
	Profile         string                `json:"profile"`
	Secrets         map[string]secretJSON `json:"secrets"`
	MissingRequired []string              `json:"missing_required"`
	MissingOptional []string              `json:"missing_optional"`
}

// parseEnvelope unmarshals a response envelope, validates the envelope-level
// fields shared by the resolve and report responses, and returns the inner
// response. kind ("resolve" or "report") labels the version-mismatch message;
// schemaOf reads the response's schema_version for the version check.
func parseEnvelope[R any](raw, kind string, expected int, schemaOf func(*R) int) (*R, error) {
	var env envelope[R]
	if err := json.Unmarshal([]byte(raw), &env); err != nil {
		return nil, err
	}
	if !env.OK {
		errKind, message := "unknown", ""
		if env.Error != nil {
			errKind, message = env.Error.Kind, env.Error.Message
		}
		return nil, &Error{Kind: errKind, Message: message}
	}
	if env.Response == nil {
		return nil, &Error{Kind: "ffi", Message: "secretspec_resolve reported ok with no response"}
	}
	if v := schemaOf(env.Response); v != expected {
		return nil, &Error{Kind: "version", Message: fmt.Sprintf(
			"unsupported %s schema version %d (expected %d); the secretspec-ffi library and this SDK are out of sync",
			kind, v, expected,
		)}
	}
	return env.Response, nil
}

// Load resolves the secrets. It returns *MissingRequiredError if a required
// secret is missing, and *Error for any other failure.
func (b *Builder) Load() (*Resolved, error) {
	if err := ensureLoaded(); err != nil {
		return nil, err
	}
	// A zero-value Builder (var b Builder; b.Load()) has a nil req, which marshals
	// to the literal `null` that serde rejects as an invalid JsonRequest. The WithX
	// setters lazily allocate req, but Load may run before any of them.
	if b.req == nil {
		b.req = map[string]any{}
	}
	payload, err := json.Marshal(b.req)
	if err != nil {
		return nil, err
	}

	raw, err := nativeResolve(string(payload))
	if err != nil {
		return nil, err
	}

	resp, err := parseEnvelope(raw, "resolve", resolveSchemaVersion, func(r *responseJSON) int { return r.SchemaVersion })
	if err != nil {
		return nil, err
	}
	if len(resp.MissingRequired) > 0 {
		return nil, &MissingRequiredError{Missing: resp.MissingRequired}
	}

	secrets := make(map[string]ResolvedSecret, len(resp.Secrets))
	for name, entry := range resp.Secrets {
		secrets[name] = ResolvedSecret{
			Value:          entry.Value,
			Path:           entry.Path,
			AsPath:         entry.AsPath,
			Source:         entry.Source,
			SourceProvider: entry.SourceProvider,
		}
	}
	return &Resolved{
		Provider:        resp.Provider,
		Profile:         resp.Profile,
		Secrets:         secrets,
		MissingOptional: resp.MissingOptional,
	}, nil
}

// reportSchemaVersion is the value-free report wire-format version this SDK
// understands; it tracks secretspec's RESOLUTION_REPORT_SCHEMA_VERSION.
const reportSchemaVersion = 1

// SecretReport is the value-free resolution outcome for one declared secret:
// how it would resolve and from where, never the value itself.
type SecretReport struct {
	Name           string
	Status         string // "resolved" | "missing_required" | "missing_optional"
	Required       bool
	SourceProvider *string
	DefaultApplied bool
	Generated      bool
	AsPath         bool
}

// Report is a value-free resolution snapshot: every declared secret and how it
// would resolve, never a value. Unlike Load, a missing required secret is
// reported as a SecretReport with Status "missing_required" rather than an
// error, so it describes a profile even when its secrets are not all available.
type Report struct {
	Provider string
	Profile  string
	Secrets  []SecretReport
}

type secretReportJSON struct {
	Name           string  `json:"name"`
	Status         string  `json:"status"`
	Required       bool    `json:"required"`
	SourceProvider *string `json:"source_provider"`
	DefaultApplied bool    `json:"default_applied"`
	Generated      bool    `json:"generated"`
	AsPath         bool    `json:"as_path"`
}

type reportResponseJSON struct {
	SchemaVersion int                `json:"schema_version"`
	Provider      string             `json:"provider"`
	Profile       string             `json:"profile"`
	Secrets       []secretReportJSON `json:"secrets"`
}

// Report resolves the value-free report (the inventory/preflight view, the same
// one the CLI exposes as `check --json`). It never returns
// *MissingRequiredError: a missing required secret appears as a SecretReport
// with Status "missing_required". It returns *Error for a genuine failure.
func (b *Builder) Report() (*Report, error) {
	if err := ensureLoaded(); err != nil {
		return nil, err
	}
	req := make(map[string]any, len(b.req)+1)
	for k, v := range b.req {
		req[k] = v
	}
	req["mode"] = "report"
	payload, err := json.Marshal(req)
	if err != nil {
		return nil, err
	}

	raw, err := nativeResolve(string(payload))
	if err != nil {
		return nil, err
	}

	resp, err := parseEnvelope(raw, "report", reportSchemaVersion, func(r *reportResponseJSON) int { return r.SchemaVersion })
	if err != nil {
		return nil, err
	}

	secrets := make([]SecretReport, len(resp.Secrets))
	for i, s := range resp.Secrets {
		secrets[i] = SecretReport(s)
	}
	return &Report{Provider: resp.Provider, Profile: resp.Profile, Secrets: secrets}, nil
}
